fdbclient/include/fdbclient/KeyBackedTypes.actor.h

```diff
@@ -0,0 +1,1083 @@
+		}
+
+		return v;
+	}
+};
+
+template <>
+inline Standalone<StringRef> TupleCodec<KeyRange>::pack(KeyRange const& val) {
+	return Tuple::makeTuple(val.begin, val.end).pack();
+}
+template <>
+inline KeyRange TupleCodec<KeyRange>::unpack(Standalone<StringRef> const& val) {
+	Tuple t = Tuple::unpack(val);
+	return KeyRangeRef(t.getString(0), t.getString(1));
+}
+
+struct NullCodec {
+	static Standalone<StringRef> pack(Standalone<StringRef> val) { return val; }
+	static Standalone<StringRef> unpack(Standalone<StringRef> val) { return val; }
+};
+
+template <class T>
+struct BinaryCodec {
+	static Standalone<StringRef> pack(T val) { return BinaryWriter::toValue<T>(val, Unversioned()); }
+	static T unpack(Standalone<StringRef> val) { return BinaryReader::fromStringRef<T>(val, Unversioned()); }
+};
+
+// Codec for using Flatbuffer compatible types via ObjectWriter/ObjectReader
+template <typename T, typename VersionOptions>
+struct ObjectCodec {
+	ObjectCodec(VersionOptions vo) : vo(vo) {}
+	VersionOptions vo;
+
+	inline Standalone<StringRef> pack(T const& val) const { return ObjectWriter::toValue<T>(val, vo); }
+	inline T unpack(Standalone<StringRef> const& val) const { return ObjectReader::fromStringRef<T>(val, vo); }
+};
+
+template <typename ResultType>
+struct KeyBackedRangeResult {
+	std::vector<ResultType> results;
+	bool more;
+
+	bool operator==(KeyBackedRangeResult const& other) const { return results == other.results && more == other.more; }
+	bool operator!=(KeyBackedRangeResult const& other) const { return !(*this == other); }
+};
+
+class WatchableTrigger {
+private:
+	Key key;
+
+public:
+	WatchableTrigger(Key k) : key(k) {}
+
+	template <class Transaction>
+	void update(Transaction tr, AddConflictRange conflict = AddConflictRange::False) {
+		std::array<uint8_t, 14> value;
+		value.fill(0);
+		tr->atomicOp(key, StringRef(value.begin(), value.size()), MutationRef::SetVersionstampedValue);
+		if (conflict) {
+			tr->addReadConflictRange(singleKeyRange(key));
+		}
+	}
+
+	template <class Transaction>
+	Future<Optional<Versionstamp>> get(Transaction tr, Snapshot snapshot = Snapshot::False) const {
+		typename transaction_future_type<Transaction, Optional<Value>>::type getFuture = tr->get(key, snapshot);
+
+		return holdWhile(
+		    getFuture,
+		    map(safeThreadFutureToFuture(getFuture), [](Optional<Value> const& val) -> Optional<Versionstamp> {
+			    if (val.present()) {
+				    return BinaryReader::fromStringRef<Versionstamp>(*val, Unversioned());
+			    }
+			    return {};
+		    }));
+	}
+
+	template <class Transaction>
+	Future<Void> watch(Transaction tr) const {
+		typename transaction_future_type<Transaction, Void>::type watchFuture = tr->watch(key);
+		return holdWhile(watchFuture, safeThreadFutureToFuture(watchFuture));
+	}
+
+// Forward declaration of this static template actor does not work correctly, so this actor is forward declared
+// in a way that works for both compiling and in IDEs.
+#if defined(NO_INTELLISENSE)
```

Comment:
This was exactly what I started with, and unfortunately since this is not a template class the class itself cannot be used within the actor definition since its definition is not complete.  So I had to move it, and then figure out how to make it work while keeping its namespace within the enclosing class.